wolfcrypt/src/random.c

```diff
@@ -800,6 +809,776 @@ int wc_RNG_TestSeed(const byte* seed, word32 seedSz)
 #endif /* HAVE_HASHDRBG */
 /* End NIST DRBG Code */
 
+#ifdef HAVE_ENTROPY_MEMUSE
+/* Define ENTROPY_MEMUSE_THREAD to force use of counter in a new thread.
+ * Only do this when high resolution timer not otherwise available.
+ */
+
+/* Number of bytes that will hold the maximum entropy bits. */
+#define MAX_ENTROPY_BYTES    (MAX_ENTROPY_BITS / 8)
+/* Number of bits stored for one sample. */
+#define ENTROPY_BITS_USED    8
+
+/* Minimum entropy from a sample. */
+#define ENTROPY_MIN          1
+/* Number of extra samples to ensure full entropy. */
+#define ENTROPY_EXTRA        64
+/* Maximum number of bytes to sample to produce max entropy. */
+#define MAX_NOISE_CNT        (MAX_ENTROPY_BITS * 8 + ENTROPY_EXTRA)
+
+/* Jitter entropy global state initialized. */
+static int entropy_memuse_initialized = 0;
+/* Global SHA-3 object used for conditioning entropy and creating noise. */
+wc_Sha3 entropyHash;
+/* Reset the health tests. */
+static void Entropy_HealthTest_Reset(void);
+
+#if !defined(ENTROPY_MEMUSE_THREAD) && \
+    (defined(__x86_64__) || defined(__i386__))
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit count of CPU cycles.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    unsigned int lo_c, hi_c;
+    __asm__ __volatile__ (
+        "rdtsc"
+            : "=a"(lo_c), "=d"(hi_c)   /* out */
+            : "a"(0)                   /* in */
+            : "%ebx", "%ecx");         /* clobber */
+    return ((word64)lo_c) | (((word64)hi_c) << 32);
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && \
+      (defined(__APPLE__) || defined(__MACH__))
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit time in nanoseconds.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    return mach_absolute_time();
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && defined(__aarch64__)
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit timer count.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    word64 cnt;
+    __asm__ __volatile__ (
+        "mrs %[cnt], cntvct_el0"
+        : [cnt] "=r"(cnt)
+        :
+        :
+    );
+    return cnt;
+}
+#elif !defined(ENTROPY_MEMUSE_THREAD) && (_POSIX_C_SOURCE >= 199309L)
+/* Get the high resolution time counter.
+ *
+ * @return  64-bit time that is the nanoseconds of current time.
+ */
+static WC_INLINE word64 Entropy_TimeHiRes(void)
+{
+    struct timespec now;
+
+    clock_gettime(CLOCK_REALTIME, &now);
+
+    return now.tv_nsec;
+}
+#elif !defined(SINGLE_THREADED) && defined(_POSIX_THREADS) && \
+      !defined(__MINGW32__)
+
+/* Start and stop thread that counts as a proxy for time counter. */
+#define ENTROPY_MEMUSE_THREADED
+
+/* Data for entropy thread. */
+typedef struct ENTROPY_THREAD_DATA {
+    /* Current counter - proxy for time. */
+    word64 counter;
+    /* Whether to stop thread. */
+    int stop;
+} ENTROPY_THREAD_DATA;
+
+/* Track whether entropy thread has been started already. */
+static int entropy_thread_started = 0;
```

Comment:
Should we protect these variables with Thread Local Storage (TLS) or THREAD_LS_T?

Reply:
No. They need to be global.
Only one thread to be started per application.
Mutex around getting entropy.